feat(cache): Adds custom s3 backend class

ERosendo opened the pull request:
This PR introduces a new Django cache backend, `S3ExpressCacheBackend`, designed to leverage AWS S3 Express One Zone for high-throughput, low-latency caching.

**Key Features:**

- **Optimized Key Management for S3:**
   
  - **Directory-like Paths:** Transforms time-based key prefixes (e.g., `N-days:key`) into directory-like structures (e.g., `N-days/key`). This strategic transformation improves S3 write throughput by distributing objects across different logical prefixes, taking advantage of S3's internal partitioning mechanisms.
 
  - **Django Versioning and Prefixes:** Seamlessly integrates with Django's existing cache versioning and key prefix mechanisms.

- **Integrated expiration timeout:**

  - **Embedded Timestamps:** Manages cache item expiration by embedding timestamps directly within the S3 object data, supporting both time-limited and persistent cache entries.
  
  - **Timeout Validation:** Enforces a critical validation rule to prevent inconsistencies: a cache item's specified timeout cannot exceed the maximum lifespan implied by its "N-days" key prefix. This ensures data integrity and predictable cache behavior.

- **Enhanced Performance:**
 
  - **Session-based Authentication:** Speeds up S3 client operations by utilizing session-based authentication for reduced authentication latency.

  - **Optimized Object Retrieval:** Implements a robust `get` method that retrieves items in chunks, checking for expiration during the retrieval process and stopping early if the item is expired. Valid data is then safely deserialized.

- Provides full implementations for standard cache operations including `set`, `get`, `add`, `has_key`, and `delete`.

ERosendo committed:
chore: Updates the `pyproject.toml` file

ERosendo committed:
feat(cache): Implement base backend class

ERosendo committed:
chore: Updates the `uv.lock` file

ERosendo committed:
feat(cache): Adds custom `set` method

ERosendo committed:
feat(cache): Implements logic for item retrieval

ERosendo committed:
feat(cache): Adds custom `has_key` method for key validation

ERosendo committed:
feat(cache): Introduces implementation of the `add` method

ERosendo committed:
feat(cache): Introduces implementation of the `delete` method

ERosendo committed:
feat(cache): Adds key path transformation logic

This commit introduces a new helper to the S3ExpressCacheBackend class to improve key management and optimize performance with S3:

- `_key_to_directory_path`: This transformation is essential for leveraging S3's internal partitioning mechanisms, which helps distribute write operations more efficiently and can significantly improve throughput for high-volume write scenarios

ERosendo committed:
feat(cache): Override and enhance make_key method

This commit overrides the `make_key` method. It extends the base class's key method by adding a step to transform the prefix of the key into a directory-like structure. This aligns object keys with best practices for distributing write operations, which helps boost throughput and performance, especially for high-volume data

ERosendo committed:
feat(cache): Introduce `parse_time_base_prefix` for extracting key duration

This commit introduces a new helper method called `parse_time_based_prefix`. This method is responsible for extracting the numeric duration (in days) from the cache key's time-based prefix, such as "N-day(s):" or "N-day(s)/".

The extracted value helps validate that a cache item's set timeout doesn't exceed the maximum lifespan indicated by its key's prefix. If the key format doesn't conform to the expected time-based prefix, a ValueError is raised.

ERosendo committed:
feat(cache): Adds timeout validation against key prefix

This commit introduces a new validation step within the `set` method to ensure that a cache item's **timeout** doesn't exceed the maximum lifespan specified by its key's time-based prefix.

ERosendo committed:
feat(cache): Introduces custom key builder function for S3 Express Backend

ERosendo committed:
tests(cache): Adds test suite for helper functions

ERosendo committed:
docs(cache): Adds docstring to the S3ExpressCacheBackend class

ERosendo committed:
perf(cache): Speeds up S3 client with session-based authentication

ERosendo committed:
feat(cache): Adds logic to handle persistent objects

ERosendo committed:
feat(cache): Enhances S3 cache test coverage

mlissner commented on django_s3_express_cache/s3_cache.py:
```diff
@@ -0,0 +1,238 @@
+        # Iterate over chunks of the S3 object's body.
+        # The first 8 bytes (chunk_size=8) are expected to be the expiration timestamp.
+        for i, chunk in enumerate(response["Body"].iter_chunks(chunk_size=8)):
+            if not i:
```
This could be more explicit like this: 

```suggestion
            if i == 0:
```

ERosendo replied:
sounds good!

mlissner commented on django_s3_express_cache/s3_cache.py:
```diff
@@ -0,0 +1,238 @@
+        # original value and return it.
+        return pickle.loads(bytes(cached_object))
+
+    def delete(self, raw_key, version=None):
```
Should this return False if the key doesn't exist? The docs are unclear on this.

ERosendo replied:
> Should this return False if the key doesn't exist? 

Yes!

You’re right about the docs 😅 . They state:

> returns True if the key was successfully deleted, False otherwise.

There’s no detail in the docs about how missing keys are handled, but I checked the implementation of the [delete](https://github.com/django/django/blob/fda3c1712a1eb7b20dfc91e6c9abae32bd64d081/django/core/cache/backends/filebased.py#L82-L90) method in one of the default cache classes, and it does return `False` when the key doesn’t exist.

ERosendo replied:
@mlissner I’ve reviewed the remaining cache class implementations. All of them return False when the key doesn’t exist. Each class uses a slightly different mechanism to achieve this:

- The [database implementation](https://github.com/django/django/blob/fda3c1712a1eb7b20dfc91e6c9abae32bd64d081/django/core/cache/backends/db.py#L215-L234) checks the number of rows deleted.

- The [Redis implementation](https://github.com/django/django/blob/fda3c1712a1eb7b20dfc91e6c9abae32bd64d081/django/core/cache/backends/redis.py#L118-L120) does something similar, converting the number of deleted keys to a boolean.

- The [locmem implementation](https://github.com/django/django/blob/fda3c1712a1eb7b20dfc91e6c9abae32bd64d081/django/core/cache/backends/locmem.py#L102-L108) relies on catching a KeyError and returns False in that case.

However, it seems the `delete_object` method from `boto3` returns a success message instead of an error when the key doesn’t exist. The boto3 docs aren’t very explicit about this, but the [AWS Java SDK docs](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3.html#deleteObject-java.lang.String-java.lang.String-) confirm this behavior for the delete operation. 

> **deleteObject**
> 
> Deletes the specified object in the specified bucket. Once deleted, the object can only be restored if versioning was enabled when the object was deleted.
> 
> If attempting to delete an object that does not exist, Amazon S3 returns a success message instead of an error message.

We could use the [head_object](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/head_object.html) method in our logic to first check whether the key exists. This method raises a `Client.exceptions.NoSuchKey` exception when the key is missing, which we can catch to return `False`. That way, our implementation would be consistent with the other cache classes.

Let me know what you think.

mlissner replied:
Sounds like the right thing to do, yeah. So if there's no key, we'd do the head and return false. If there is a key, we'd do the head, find the key exists, do the delete, and return true.

mlissner reviewed (commented):
This is exciting to see and I haven't read the tests, but I did a quick review of the cache, and I have a few notes for the next steps of this. If you can complete these, I think we'll be good to get another review somebody else on the team. (Right now, I think the full design/architecture is a bit hard to understand without reading the issue description, but this should stand alone.)

So with that said, congrats on getting it out, and some thoughts:

1. Why use pickle? I worry about a few things:
    - Is it secure? I keep hearing about how pickle can run arbitrary code. If we use this, do we need to warn about the cache's ability to run code? 
    - Is it fast? I don't know how fast pickle operates on large objects. Worth considering? 

    I also wonder if minor compression using zip is useful for speed. It's often the case with slow read/write speeds that compressing then (un)storing is faster than doing it without compression. Comes with a CPU tradeoff though, of course. Maybe this fits into future work, but this *does* imply that we might want compression type as a parameter in our byte format. So perhaps it goes:

    | Bytes 0-8 | Bytes 9-x | Bytes x-y | Bytes y-z | Bytes z-* |
    |--------|--------|--------|--------|---|
    | expiration date | compression technique | version | extra spacer | data |
    
    Note that I added the version in there. I'm guessing this will be necessary for future upgrades of the system, to prevent crashes with unexpected versions. Not sure. Seems useful. Extra can be a blank spot for the unexpected. TCP has this. We might need it someday, so let's include it, since it's easy?

    The above might not be the right ordering for this. Assuming you like this idea, I think you'll know once you start coding it up.

2. What cache methods are missing? We should at least document them.

3. On the subject of documentation, I'd love a readme at this stage to explain:
    - Why we made this
    - How it's better or worse than other options
    - The design of it (things like struct, pickle, early abortions, key design, etc — the things we discussed in the issue).
    - Why this is better/worse than the similar options.
    - The fact that we have this in prod in CL.
    - Its installation, set up, and configuration (including how to make the lifecycle rules?)

    This readme should show up in the python packaging when we do that.

-----

Future work:

1. We need automatic deployment for this, perhaps like we have for Juriscraper or the other python packages we host.

2. We need the changelog tooling we've got elsewhere to ensure we always have an entry for each PR.

3. I saw that it only supports >Django 5.0, but I imagine it'd work going back quite a ways. I guess we can just open an issue for this and let others deal with adding a testing matrix and such, but I think we have examples of testing matrices in at least one of our tools using tox if we want to link to it or go for it.

----

I think that's it for now. This looks great!

ERosendo committed:
refactor(cache): Uses explicit zero check for clarity

ERosendo commented:
@mlissner Here are my responses to a few of the questions:

> Why use pickle?

When I started working on this, I reviewed the core cache backends in the Django repository. I noticed that both the file-based and database-backed implementations use pickle. Each with its own custom write method to handle their specific data format, but both relying on pickle under the hood. I felt it made sense to use something already proven and compatible with our use case. 

> Is it secure? I keep hearing about how pickle can run arbitrary code. If we use this, do we need to warn about the cache's ability to run code?

I understand the concern. the Python pickle documentation warns about the risk of arbitrary code execution right in the first paragraph. In our case, the cache only stores and retrieves data we create ourselves, with no possibility for untrusted input to be injected, so I believe the security risk for us is low.

The documentation suggests considering safer serialization formats such as [JSON](https://docs.python.org/3/library/json.html#module-json), but JSON is limited to basic types, and our goal is to create a backend as flexible as Django’s existing ones, capable of caching complex objects like templates and query results, while using a faster storage engine.

These type limitations also rule out faster JSON variants like [orjson](https://github.com/ijl/orjson) and [ujson](https://github.com/ultrajson/ultrajson). I also tested [msgpack](https://github.com/msgpack/msgpack-python) (safer than pickle but also a data-only format), but it failed to serialize certain types.

That said, I can continue exploring alternative formats that might offer a better balance of safety and compatibility.

> Is it fast? I don't know how fast pickle operates on large objects. Worth considering?

Yes, speed is worth considering. According to the pickle module documentation:

- Python 3.4 introduced support for very large objects, the ability to pickle more object types, and several data format optimizations.

- Python 3.8 added support for out-of-band data (reducing memory overhead for large objects) and improved the speed of in-band data serialization.

With these improvements, modern pickle (protocol ≥ 4) is generally fast enough for small to medium-sized Python objects.

> I also wonder if minor compression using zip is useful for speed.

I think this makes sense, and this backend wouldn’t be the first to add compression to stored objects. In fact, the file-based cache backend uses zlib to [compress the pickled data](https://github.com/django/django/blob/fda3c1712a1eb7b20dfc91e6c9abae32bd64d081/django/core/cache/backends/filebased.py#L42-L45) when writing cache files.

> this does imply that we might want compression type as a parameter in our byte format.
> 
> so let's include it, since it's easy?

I like this idea. Including extra metadata in the header now will make the format more flexible and easier to extend in the future.

> What cache methods are missing? We should at least document them.

The methods not included in this PR are [touch](https://docs.djangoproject.com/en/5.2/topics/cache/#django.core.cache.cache.touch) and [clear](https://docs.djangoproject.com/en/5.2/topics/cache/#django.core.cache.cache.clear). This PR also doesn’t explicitly define `incr` or `decr`, but the base class already provides default implementations for these using the existing methods.

> On the subject of documentation, I'd love a readme at this stage

sure! 🫡

mlissner commented:
Pickle seems fine for now, thanks. 

Let's tweak to include versioning and the option for compression in the header. 

The rest sounds great, thank you!

ERosendo committed:
feat(cache): Checks if S3 object exists before deleting

Adds a `head_object` check to ensure the key exists before attempting deletion. Returns `False` when the key is missing instead of raising an error.

ERosendo committed:
feat(cache): Makes pickle protocol configurable across cache backend

ERosendo committed:
feat(cache): Move code into __init__.py to simplify imports

Relocated functionality from a separate file into __init__.py to avoid extra path components when importing, making usage cleaner and
straightforward.

ERosendo committed:
feat(cache): Lazily load boto3 client with cached_property

Use cached_property to defer boto3 client initialization until first access, reducing startup overhead and avoiding unnecessary resource usage.

ERosendo committed:
feat(cache): Optimizes has_key by requesting only expiration timestamp

ERosendo committed:
feat(cache): Adds custom header for entries with versioning and compression

- Introduces header format for cached objects:
  - 8 bytes: expiration timestamp (ns)
  - 2 bytes: format version
  - 2 byte : compression type (0 = none, 1 = zlib, etc.)
  - 8 bytes: reserved extra space (future use)
- Added helper functions `make_header` and `parse_header` for packing/unpacking
- Updated `set` method to prepend header before serialized data
- Future-proofs cache format for version upgrades and optional compression

ERosendo committed:
feat(docs): Updates README to reflect project details

ERosendo committed:
feat(docs): Add note about missing touch and clear methods

ERosendo commented:
Hi @mlissner 👋 

The README is now ready, and I’ve addressed the comments you shared via email. Here’s a summary of the changes and key decisions:

- **Header improvements:** introduced a new compact binary header to store **expiration**, **versioning**, optional **compression**, and reserved extra space for future metadata. 

- Commit [7f97212](https://github.com/freelawproject/django-s3-express-cache/pull/1/commits/7f97212df279f14601aa39bae17c75e14b25ce40) moved the code into __init__.py to simplify the import path.

- In [3109ea5](https://github.com/freelawproject/django-s3-express-cache/pull/1/commits/3109ea5483d8fa3265a59051b170a6c9df9c8820), the backend class was refactored to lazily load the s3 client using a @cached_property, reducing startup overhead.

- [a3983c4](https://github.com/freelawproject/django-s3-express-cache/pull/1/commits/a3983c48667b2b362a783e638ae3be3f150fa6e8) improves the` has_key()` method. It now uses a [Range](https://www.rfc-editor.org/rfc/rfc9110.html#name-range) request to fetch only the header, allowing expiration checks without downloading the full object.
 
  The email comment suggested using `head_object` alone, but in our case that wouldn’t work reliably. Since we delegate cleanup to S3 lifecycle rules, there can be a delay between an item’s expiration and its actual deletion. If we relied on `head_object` only, an expired object that still exists would be treated as valid. Other implementations that clean more frequently or precisely could safely use `head_object`, but for our workflow, the header check ensures correctness while minimizing unnecessary data transfer. 

- The `add()` method retains the check for expiration before writing, as mere existence check isn’t sufficient. Adam suggested attempting to save without overwriting (e.g., using the `If-None-Match` header), but that wouldn’t work in our workflow due to the potential delay between expiration and deletion discussed above.

The README now reflects these design choices, explains serialization, key structure, header format, lifecycle rules, and trade-offs.

mlissner commented on README.md:
Sorry this took me a little while. I wanted to try my hand at improving the readme, and I wanted to ask ChatGPT what was missing from it. Below is the result (TLDR, it was really good and ChatGPT didn't come up with much). 

Want to give this a read, fill in the missing bits, fix anything that you think could be better, and commit it? 

# django-s3-express-cache

**A scalable, open source Django cache backend powered by [Amazon S3 Express One Zone][1z] — cheaper, durable, and ready for production.**


## Why?

Django ships with two main cache backends, but neither is a great fit for many or large objects:

| Backend            | Pros                                    | Cons                                                                 |
|--------------------|-----------------------------------------|----------------------------------------------------------------------|
| **Database cache** | Easy to set up                          | Does a `COUNT(*)` on every `get`, `set`, or `touch`, which [does not perform][1] on large cache tables. |
| **Redis/Memcached**| Fast, widely used                       | Expensive to run at scale (large RAM bills, cluster management)      |

On the other hand, [S3 Express One Zone][1z] provides an S3 bucket with single-digit-millisecond latency that is cheap, durable, and can scale to millions of objects, large and small. 

S3 Express does not support automatic item expiration, so we use [S3 lifecycle rules][lifecycle], a fixed-width header prepended to each item, and clever key names to manage and cull the cache as needed.


## Features

- 🚀 **Scalable & cost-effective** — cache huge datasets without memory overhead.
- 🧹 **Automatic background cleanup** — expired keys are culled automatically with [S3 Lifecycle Rules][lifecycle].
- ⚡ **Fast reads** — header-only [range requests][rr] detect expired objects without fetching full objects.
- 🔮 **Future-proof format** — compact binary header with versioning and reserved fields inspired by TCP frames for future functionality.
- 🛠 **Easy integration** — configure in Django `CACHES`, add lifecycle rules, and go.


## Requirements

- **Django 5.x**  
- **Python ≥ 3.13**  
- **boto3 v1.38.36+**  
- Works in any AWS region where S3 Express One Zone is available
- Best used in the same Availability Zone as your application  


## Design

### Motivation

This backend was inspired by an issue raised in [CourtListener’s repository](https://github.com/freelawproject/courtlistener/issues/5304). In short:

- Django’s DB cache [can become a performance bottleneck under heavy load][1], especially when culling expired rows. Queries like `SELECT COUNT(*) FROM django_cache` cause significant slowdowns once the cache table grows large. In our experience running CourtListener, the DB cache is one of the heaviest consumers of database resources.

- Django's in-memory caches do not scale well when caching large objects or many small ones.

- S3 is highly scalable, cost-effective, and capable of storing large objects. Instead of relying on costly culling queries (like the DB cache), we can use [S3 lifecycle rules][lifecycle] to automatically cull stale entries, keeping performance stable without scripts or app-level logic.

This implementation builds on those ideas and delivers a production-ready, efficient, and extensible cache backend, designed to integrate naturally with Django’s caching framework.


### Cache Key Name Requires for Maximal S3 Throughput and Automatic Culling

- S3 Express One Zone uses [directory buckets][dirbuck], which support [Lifecycle policies][lifecycle] but only with limited filters (prefix and size, no tags). 
 
    To align with these constraints, our design requires explicit time-based key prefixes (e.g., `1-days/`, `7-days/`, `30-days/`) that align with the expiration of your item. Any expiration is supported up to 1,000 days, and your cache key must correspond to the number of days greater than your items expiration. For example, an item expiring today should have a key like `1-days:foo` and an item expiring in 25 hours should have a key like `2-days:bar`.

    This ensures cache entries can be automatically culled using prefix-based lifecycle rules.

- Keys of the form `N-days:actual_key` are added to the bucket as `N-days/actual_key` (with a slash instead of a colon). This spreads objects across S3 key prefixes, improving S3 partitioning and request throughput.

- When adding something to the cache, the key name is validated against the expiration date for the item. If the expiration exceeds the `N-days` limit, the write is rejected. This prevents accidentally storing long-lived items under a short-lived namespace and keeps lifecycle-based culling predictable. Such errors will generally be caught during development.

[dirbuck]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/directory-buckets-overview.html

### Performance Optimizations

To optimize data transfer and improve performance, the backend implements early exits:

- **`has_key`**
  
  Uses an [S3 `Range` request][rr] to fetch only the relevant 20 bytes from the header (see header format below for details).  
  - If the item is expired → treated as a cache miss without downloading the full value → return `False`.
  - If the item is persistent or still valid → considered a hit → return `True`.

- **`get`**

  Streams the object in header-sized chunks.  

  After reading the header (first chunk), expiry is evaluated.  
   
  - If expired → the operation exits immediately without fetching the remaining data.  
  - If valid → streaming continues to reconstruct the cached object.


### Header format (fixed-width, versioned)

We prepend a compact header to every object stored in the cache. The current layout of the header is ([struct](https://docs.python.org/3/library/struct.html) format: dHHQ):

| Field           | [Type](https://docs.python.org/3/library/struct.html#format-characters) | Bytes | Notes                                                             |
|-----------------|------|-------|-------------------------------------------------------------------|
| expiration_time | d    | 8     | UNIX timestamp in seconds (float). `0` means persistent.           |
| header_version  | H    | 2     | Starts at `1`. Used for compatibility checks and future extensibility. |
| compression_type| H    | 2     | `0 = none`. Reserved for future use (e.g., zlib, zstd).           |
| extra (reserved)| Q    | 8     | Reserved for future metadata                                      |

Using a fixed-width header allows the cache to [Range-read][rr] only the header. Items remain in the cache until [S3 Lifecyle rules][lifecycle] complete, so this allows your application to check the expiration of an object before downloading it. If the item is expired, that's a cache miss. If not, the entire object is downloaded and returned.

> [!NOTE]
> The code is written to treat mismatched header versions as unsupported (safe default). You can add backward parsers in the future if needed.


### Lazy boto3 Client Initialization

Creating a boto3 client (and even importing boto3 itself) can be relatively expensive. To avoid adding this overhead to Django’s general startup time, the backend initializes the client **lazily** using a `@cached_property`.  

This means:
- The boto3 client is created only on first use.  
- Subsequent accesses reuse the cached client instance.  
- Application startup remains fast, while still ensuring efficient reuse of the client once it’s needed.


### Security

This backend uses Python’s `pickle` with `HIGHEST_PROTOCOL`, providing fast serialization and broad support for Python object types.

Django’s own [file-based](https://github.com/django/django/blob/8b229b4dbb6db08428348aeb7e5a536b64cf8ed8/django/core/cache/backends/filebased.py) and [database-backed](https://github.com/django/django/blob/8b229b4dbb6db08428348aeb7e5a536b64cf8ed8/django/core/cache/backends/db.py) cache backends both rely on pickle internally, each with their own write method. We chose to follow this pattern for consistency, compatibility, and flexibility—especially since our goal was a backend as capable as Django’s built-ins.

> [!CAUTION]
> Pickle should only be used with trusted data that your own application writes and reads. Never unpickle untrusted payloads. If your use case requires stricter, data-only serialization, formats like JSON or MessagePack are safer but keep in mind their type limitations.


---

## Usage

There are five steps to using this cache:

1. Install it

2. Configure it in your django settings

3. Set up the S3 Express bucket

4. Configure lifecycle rules for automatic cache culling

5. Use it! 

### Installation

From PyPI:

```sh
pip install django-s3-express-cache
```

From GitHub (latest dev):

```
pip install git+https://github.com/freelawproject/django-s3-express-cache.git@master
```


### Configuration

We do not recommend this cache as your primary, default cache. Instead, it should be used as a secondary cache for larger or longer-living objects by putting something like the following in your Django settings:

```
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": REDIS_URL,
    },
    "s3": {
        "BACKEND": "django-s3-express-cache.S3ExpressCacheBackend",
        "LOCATION": S3_CACHE_BUCKET_NAME,
    }
}
```

### Bucket Set Up

You must use an S3 Express One Zone (Directory bucket).

👉 TODO: Developer to add aws s3api create-bucket example here.


### Lifecycle Rule Set Up

A timestamp stored in the item's fixed-width header is used to ensure that items expire at the correct time. 

Lifecycle rules are used to cull stale items from the cache. Rules should be configured to cull objects by prefix.

For example:
- Objects under 7-days/ expire after 7 days
- Objects under 30-days/ expire after 30 days

```json
{
  "Rules": [
    {
      "ID": "Expire-7-days-prefix",      ①
      "Filter": { "Prefix": "7-days/" }, ②
      "Status": "Enabled",               ③ 
      "Expiration": { "Days": 7 }        ④
    },
    {
      "ID": "Expire-30-days-prefix",
      "Filter": { "Prefix": "30-days/" },
      "Status": "Enabled",
      "Expiration": { "Days": 30 }
    }
  ]
}
```

① Give the rule a name
② Set the rule to the "7-days" directory
③ Enable the rule
④ Set the expiration time to match the directory name

These lifecycle rules complement the cache’s in-object header expiration. The header allows our implementation to short-circuit reads (treating expired items as misses), while S3 lifecycle policies ensure expired data is culled from the bucket.

The following script demonstrates how to configure up to 1,000 lifecycle rules in a bucket.
To run it, your IAM must have at least the following permissions:

- `s3:PutLifecycleConfiguratio`
- `s3:GetLifecycleConfiguration`


```python
import boto3

# Replace with your bucket name
BUCKET_NAME = "your-bucket-name"

s3 = boto3.client("s3")

rules = []
for i in range(1, 1000):
    # Handle pluralization
    suffix = "days" if i > 1 else "day"
    prefix = f"{i}-{suffix}"

    rules.append({
        "ID": f"expire-{i}-{suffix}",
        "Filter": {"Prefix": prefix},
        "Status": "Enabled",
        "Expiration": {"Days": i},
    })

lifecycle_config = {"Rules": rules}

response = s3.put_bucket_lifecycle_configuration(
    Bucket=BUCKET_NAME,
    LifecycleConfiguration=lifecycle_config
)
```

### Use It!

👉 TODO: Can you provide some usage examples and explanations here? 


## Roadmap

 - Reserved header fields allow future compression support (zlib/zstd).
 - clear() and touch() methods open for contribution.
 - Performance benchmarks welcome.


## License

This project is licensed under the permissive BSD license.
Pull requests and contributions are welcome!

## Acknowledgements

Inspired by [CourtListener issue #5304](https://github.com/freelawproject/courtlistener/issues/5304) and [Django issue 32785][1].

[1z]: https://aws.amazon.com/s3/storage-classes/express-one-zone/
[lifecycle]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html
[rr]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/Range_requests
[1]: https://code.djangoproject.com/ticket/32785

mlissner reviewed (approved):
I think we're on our way here. I did some work on the readme. Once it's incorporated, let's have @albertisfu take a look!

ERosendo committed:
fix(cache): Use time.time_ns for expiration check in S3ExpressCacheBackend

ERosendo committed:
feat(cache): Updates header to use unsigned long long for expiration time

ERosendo committed:
feat(docs): Improves README clarity on features and usage

ERosendo commented:
@albertisfu This PR is ready for you! 🚀

albertisfu commented on README.md:
```diff
@@ -1,94 +1,365 @@
+        "LOCATION": REDIS_URL,
+    },
+    "s3": {
+        "BACKEND": "django-s3-express-cache.S3ExpressCacheBackend",
```
I tested this in CL and noticed that the module name must be the following for it to work:

```suggestion
        "BACKEND": "django_s3_express_cache.S3ExpressCacheBackend",
```

ERosendo replied:
Thanks for catching this!

albertisfu commented on django_s3_express_cache/__init__.py:
```diff
@@ -0,0 +1,310 @@
+            prefix, or the original key if no transformation is necessary or
+            applicable.
+    """
+    pattern_with_colon = r"^(^\d+-days?):(.*)$"
```
Is it the double `^` required?

Similar comment for `parse_time_base_prefix`

ERosendo replied:
oh no, It's redundant. Thanks for the catch, I'll go fix those regexes now!

albertisfu commented on django_s3_express_cache/__init__.py:
```diff
@@ -0,0 +1,310 @@
+        Build a binary header for cache storage.
+
+        Layout:
+        - Bytes 0-7   : expiration time (float, seconds since epoch in ns)
```
Isn't the expiration time an integer rather than a float?

ERosendo replied:
You're right!

albertisfu commented on django_s3_express_cache/__init__.py:
```diff
@@ -0,0 +1,310 @@
+            0,
+        )
+
+    def parse_header(self, header_bytes) -> tuple[float, int, int, int]:
```
Consider adding type hints to class method arguments.

ERosendo replied:
done! ✅

albertisfu commented on django_s3_express_cache/__init__.py:
```diff
@@ -0,0 +1,310 @@
+        if version != self.header_version:
+            raise ValueError(f"Unsupported cache entry version: {version}")
+
+        # If expiration_timestamp is 0, it's a persistent object.
```
According to Django cache conventions:

https://docs.djangoproject.com/en/5.2/topics/cache/#cache-arguments

> TIMEOUT to None so that, by default, cache keys never expire. A value of 0 causes keys to immediately expire (effectively “don’t cache”).


Should we follow the same convention? So 0 is an object that shouldn't be cached, and None a persistent object?

ERosendo replied:
Good point 👍. I think it makes sense to follow Django’s convention here for consistency with the ecosystem. Thanks

albertisfu commented on django_s3_express_cache/__init__.py:
```diff
@@ -0,0 +1,310 @@
+                    continue
+                # If the current time is past the expiration, the item is expired,
+                # so return the default value.
+                if time.time_ns() > expiration_timestamp:
```
Would it be a good idea to delete the cached object if we detect it has expired, or is it better for performance to let the S3 lifecycle policy handle it?

ERosendo replied:
> Would it be a good idea to delete the cached object if we detect it has expired, or is it better for performance to let the S3 lifecycle policy handle it?

Good question. In practice, it’s better not to delete expired objects on read, and to rely on the S3 lifecycle rules for cleanup. I believe this is the best balance because it offloads the work from the application, making read operations faster while also being cheaper and simpler.

albertisfu commented on django_s3_express_cache/__init__.py:
```diff
@@ -0,0 +1,310 @@
+
+    match = re.match(pattern, key)
+    if not match:
+        raise ValueError("Key does not have a valid time prefix")
```
This error is raised if the user sets a `KEY_PREFIX` in the cache settings e.g: `"KEY_PREFIX": "s3-cache"`

Since the key becomes something like:
`s3-cache/1-day/new_key_1`

I think we should either handle this case where the user adds a KEY_PREFIX, or block it and simply document that it's not supported.

ERosendo replied:
Thanks for catching this. I'll refine the logic to consider the prefix when it's configured

albertisfu reviewed (changes requested):
Thanks, @ERosendo. The implementation looks great, and it's working as expected so far. 

I just found a couple of small issues and have some suggestions.

ERosendo committed:
fix(docs): Corrects backend import path in README example

ERosendo committed:
fix(cache): Corrects regex patterns in key transformation helpers

ERosendo committed:
fix(cache): Updates expiration_time parameter type hint

ERosendo committed:
feat(cache): Adds missing type hints to S3ExpressCacheBackend

ERosendo committed:
feat(cache): support optional `key_prefix` in parse_time_base_prefix

The function now accepts a `key_prefix` parameter, which is stripped before parsing the time-based component of the cache key. This allows the parser to handle keys with consistent prefixes in addition to the "<N>-day(s):" or "<N>-day(s)/" formats.

ERosendo committed:
feat(cache): Aligns timeout handling with Django conventions

- Updated `set()` to follow Django cache semantics:
  - `timeout=None` → persistent object (never expires).
  - `timeout=0` → do not cache (skip writing to S3).

ERosendo committed:
test(cache): Adds coverage for timeout=0 case

This commit introduces a unit test to ensure that setting a key with `timeout=0` skips caching and does not call `put_object` on S3. This confirms the backend follows Django’s convention where `timeout=0` means “don’t cache."

ERosendo commented:
@albertisfu thank you for your feedback. I've addressed the comments, and this PR is ready for another look.

albertisfu commented on README.md:
```diff
@@ -1,94 +1,365 @@
+  "Rules": [
+    {
+      "ID": "Expire-7-days-prefix",      (1)
+      "Filter": { "Prefix": "7-days/" }, (2)
```
Will this filter prefix expiration rule work if the user sets a `KEY_PREFIX` in the settings?

So the cached object’s path becomes:
` s3-cache/2-days:path:to:item`

ERosendo replied:
> Will this filter prefix expiration rule work if the user sets a KEY_PREFIX in the settings?

No, if a `KEY_PREFIX` is configured in Django’s cache settings, the lifecycle filter rules from the readme won’t match because they don’t account for the prefix. To make lifecycle cleanup work correctly, the prefix needs to be included when provisioning the rules. In other words, the lifecycle rule should target:

```
<KEY_PREFIX>/<time-based-prefix>/
```

albertisfu replied:
In that case, would it be worth updating the documentation to let users know that when using a `KEY_PREFIX`, the rules should change?

ERosendo replied:
I'll update the readme

ERosendo replied:
done 👍

albertisfu commented on django_s3_express_cache/__init__.py:
```diff
@@ -0,0 +1,341 @@
+        """
+        if timeout == DEFAULT_TIMEOUT:
+            timeout = self.default_timeout
+        # The key will be made persistent if None used as a timeout.
```
I was wondering what the right approach would be if a user does something like:

`client.set("1-days:example-key-prefix_2", {"foo": "bar-prefix"}, timeout=None)`

That would mean the object is supposed to be persistent, but it’s stored with a 1-days prefix, so it will be removed after one day. Right?

Should we have a prefix rule like `"persistent"` so those objects don’t expire?
Also, maybe we need a validation that raises an error similar to the one in https://github.com/freelawproject/django-s3-express-cache/blob/79503daf1b3495b849414de78e680371b5ff08fb/django_s3_express_cache/__init__.py#L220

so a persistent key is not stored with the wrong prefix.

ERosendo replied:
> That would mean the object is supposed to be persistent, but it’s stored with a 1-days prefix, so it will be removed after one day. Right?

You're absolutely right. I’ll update the code to raise an exception if a permanent key is used with a time-based prefix.

albertisfu replied:
Great! Can we document in the `README` the correct approach to caching a persistent object, for example, by using a non-day prefix? Maybe we could also add an example.

ERosendo replied:
Sure!

ERosendo replied:
Done 👍

albertisfu commented on django_s3_express_cache/__init__.py:
```diff
@@ -0,0 +1,341 @@
+        if version != self.header_version:
+            raise ValueError(f"Unsupported cache entry version: {version}")
+
+        # If expiration_timestamp is 0, it's a persistent object.
```
This comment seems outdated.

ERosendo replied:
Not exactly. this `expiration_timestamp` isn’t the value provided by the user. t’s parsed from the header stored in S3. Since we store the expiration as an integer, we can’t use `None`, so I chose a different falsy value to represent persistent records when a user sets a key to never expire.

albertisfu replied:
oh you're right!

albertisfu reviewed (commented):
Thanks @ERosendo for the changes. They look great. Just a couple of additional comments, and I think we’ll be ready to merge this.

ERosendo committed:
fix(cache): Refines set() to handle persistent objects with time-based prefixes

ERosendo committed:
test(cache): reject persistent keys with time-based prefixes

ERosendo committed:
docs(README): add section on persistent objects and usage examples

ERosendo committed:
docs(README): clarify usage of KEY_PREFIX in S3 Lifecycle rules

ERosendo commented:
Thanks @albertisfu, I've added your suggestions to the README file

albertisfu reviewed (approved):
Thank you! @ERosendo this looks now great and ready to merge!

:shipit:

mlissner commented:
Happy Friday!!!
